llvm/lib/Transforms/Utils/LoopMetadata.cpp

```diff
@@ -0,0 +1,235 @@
+//===-- LoopMetadata.cpp - Convert Loop Metadata to assumes --*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopMetadata.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+
+#define DEBUG_TYPE "loop-metadata"
+
+using namespace llvm;
+
+/// extract the step size by which the loop IV changes.
+int getSCEVStepSize(const SCEV *S);
+
+PreservedAnalyses LoopMetadata::run(Loop &L, LoopAnalysisManager &AM,
+                                    LoopStandardAnalysisResults &AR,
+                                    LPMUpdater &U) {
+  SE = &AR.SE;
+  AC = &AR.AC;
+  DT = &AR.DT;
+
+  assignLoopMetadata(L);
+
+  return PreservedAnalyses::all();
+}
+
+bool LoopMetadata::assignLoopMetadata(Loop &L) {
+  this->L = &L;
+  Context = &L.getHeader()->getParent()->getContext();
+
+  std::optional<int> MinIterCount = getMinTripCount(&L);
+
+  // dump loop summary
+  LLVM_DEBUG(if (L.getLoopPreheader()) {
+    dbgs() << "Preheader:" << L.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << L.getHeader()->getName() << "\n");
+
+  if (MinIterCount.has_value() && MinIterCount.value() > 0) {
+    this->MinIterCount = MinIterCount.value();
+
+    LLVM_DEBUG(L.getHeader()->getParent()->dump(););
+    addAssumeToLoopHeader();
+    return true;
+  }
+
+  return false;
+}
+bool LoopMetadata::canExtractIncrement(const SCEV *S) {
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  if (!AR)
+    return false;
+
+  const SCEVConstant *SCEVConst =
+      dyn_cast<SCEVConstant>(AR->getStepRecurrence(*SE));
+  if (!SCEVConst) {
+    return false;
+  }
+  LoopStepSize = getSCEVStepSize(SCEVConst);
+  IsLoopIncrementing = LoopStepSize > 0;
+  return true;
+}
+
+Value *LoopMetadata::calcMinIterValue(const SCEV *S, int MinIterCount,
+                                      LLVMContext *Context) {
+  assert(MinIterCount > 0);
+
+  // extract loop counter increment/decrement
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  if (!AR) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: SCEV S i not a SCEVAddRecExpr ";
+               S->dump());
+    return nullptr;
+  }
+  const SCEV *ConstExpr = AR->getOperand(1);
+  if (!AR->isAffine() || !isa<SCEVConstant>(ConstExpr)) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: Unknown SCEVAddRecExpr ";
+               AR->dump());
+    return nullptr;
+  }
+
+  int MinIterValue = std::abs(LoopStepSize * MinIterCount);
+  // calculate the minimum iteration value, since SGE is used, subtract 1
+  MinIterValue--;
```

Comment:
I'd rather not have used the abs. I think that MinIterCount gives us that the original loop condition is still true after MinItercount iterations. We can substitute InitValue + MinIterCount * LoopStepSize) for the original AddRecExpr and reason that it is true. Then we don't need to know the direction to compute new boundaries, or worry about using SGE.
What we do need to know it whether the condition represents the exit or the loopedge so that we can properly invert the condition. It is fine to put in a limitation if this is (almost) always the case, but we do need to test it.